ray-operator/controllers/ray/utils/util.go

```diff
@@ -71,6 +71,40 @@ func IsCreated(pod *corev1.Pod) bool {
 	return pod.Status.Phase != ""
 }
 
+func FindHeadPodReadyCondition(pods corev1.PodList) metav1.Condition {
+	var headPod *corev1.Pod
+	for _, pod := range pods.Items {
+		if pod.Labels[RayNodeTypeLabelKey] == string(rayv1.HeadNode) {
+			headPod = &pod
+			break
+		}
+	}
+
+	replicaHeadReadyCondition := metav1.Condition{
+		Type:   string(rayv1.HeadReady),
+		Status: metav1.ConditionFalse,
+		Reason: "HeadPodNotReady",
```

Comment:
This reason is not particular useful because it's already implied with the `HeadReady` condition being `False`.  Can the reason be more specific, like `PodCrashLoopBack`,  `ContainerCreating`, `Pending`, etc?

Reply:
I hope to make the condition consistent with Pod's [PodReady](https://pkg.go.dev/k8s.io/api/core/v1#PodConditionType) condition which both `Reason` and `Message` are empty. However, @cchen777 found that `metav1.Condition` has a limitation where `Reason` should not be empty, which is different from [PodCondition](https://pkg.go.dev/k8s.io/api/core/v1#PodCondition). See https://github.com/ray-project/kuberay/pull/2261#discussion_r1687403226 for more details.

We can't keep `Reason` and `Message` empty due to `metav1.Condition`'s limitation, but I still want to make our `HeadReady` condition closer to `PodReady`. Hence, I may prefer to keep the `Reason` and `Condition` simple ("HeadPodNotReady" is OK for me). We can add more details, such as `CrashLoopBack`, `ContainerCreating`, and `Pending`, when we observe enough use cases in the future.

I am curious whether there are obvious use cases for details about whether the Pod is ready or not. I assume that if there are obvious use cases, the upstream `PodCondition` should not leave `Reason` and `Message` empty.

Reply:
> pod status condition's reason and message are empty all the time
apology that this statement i mentioned was actually not true, `Reason` is empty when pod is ready, but when pod is not ready yet, there's actually [non empty reasons](https://github.com/kubernetes/kubernetes/blob/c75e30d0495dc42a017bc1f4565831ce4d1c928a/pkg/kubelet/status/generate.go#L32-L42), i think we can propagate reason from Pod itself

Reply:
The reason being empty or something like `HeadPodReady` when the pod is ready is fine I think. But when it's not ready, the reason should never be empty because we have information from the Pod for why it's not ready (failing readiness probe, crashing, OOMkill, etc)

Reply:
update the logic to use https://github.com/ray-project/kuberay/blob/41bd563f3af567efed34bd3181957f1005e842eb/ray-operator/controllers/ray/utils/util.go#L82-L96 so that when pod is not ready yet we can get the more accurate reason 
```
kubectl get raycluster rayservice-sample-raycluster-tnm5m -o json | jq .status
{
  "conditions": [
    {
      "lastTransitionTime": "2024-07-25T04:25:32Z",
      "message": "containers with unready status: [ray-head]",
      "reason": "ContainersNotReady",
      "status": "False",
      "type": "Ready"
    }
  ],
  "desiredCPU": "2500m",
  "desiredGPU": "0",
  "desiredMemory": "4Gi",
  "desiredTPU": "0",
  "desiredWorkerReplicas": 1,
  "endpoints": {
    "client": "10001",
    "dashboard": "8265",
    "gcs-server": "6379",
    "metrics": "8080",
    "serve": "8000"
  },
  "head": {
    "podIP": "10.244.0.9",
    "podName": "rayservice-sample-raycluster-tnm5m-head-m8h95",
    "serviceIP": "10.244.0.9",
    "serviceName": "rayservice-sample-raycluster-tnm5m-head-svc"
  },
  "lastUpdateTime": "2024-07-25T04:25:54Z",
  "maxWorkerReplicas": 5,
  "minWorkerReplicas": 1,
  "observedGeneration": 1
}
```